File: app/schemas/source/postgresql.py
from typing import Literal, Self

from pydantic import BaseModel, model_validator

from app.schemas.source.types import SQLSourceBinaryHandlingMode


class PostgreSQLSourceCreateConfig(BaseModel):
    connector: Literal["postgresql"] = "postgresql"
    host: str
    port: int = 5432
    username: str
    password: str
    database: str
    snapshot_table_schema: str = "public"
    slot_name: str
    publication_name: str
    binary_handling_mode: SQLSourceBinaryHandlingMode = (
        SQLSourceBinaryHandlingMode.BYTES
    )
    heartbeat_enabled: bool = False
    heartbeat_interval: int | None = None
    heartbeat_schema: str | None = None
    heartbeat_table: str | None = None

    # {
    #     "schema1": {
    #         "table1": ["column1", "column2"],
    #         "table2": ["column1", "column2"],
    #     },
    #     "schema2": {
    #         "table1": ["column1", "column2"],
    #         "table2": ["column1", "column2"],
    #     },
    # }
    table_hierarchy: dict[str, dict[str, list[str]]]

    @model_validator(mode="after")
    def validate_heartbeat_fields(self) -> Self:
        if self.heartbeat_enabled:
            if self.heartbeat_interval is None:
                raise ValueError(
                    "Heartbeat interval is required when heartbeat is enabled"
                )
            if self.heartbeat_schema is None:
                raise ValueError(
                    "Heartbeat schema is required when heartbeat is enabled"
                )
            if self.heartbeat_table is None:
                raise ValueError(
                    "Heartbeat table is required when heartbeat is enabled"
                )
        return self

File: internal/schemas/source/postgresql.go
package schemas

// PostgreSQLSourceCreateConfig represents the configuration for a PostgreSQL source.
type PostgreSQLSourceCreateConfig struct {
	Host                string                         `json:"host" validate:"required,hostname"`
	Port                *int                           `json:"port"`
	Username            string                         `json:"username" validate:"required"`
	Password            string                         `json:"password" validate:"required"`
	Database            string                         `json:"database" validate:"required"`
	SnapshotTableSchema *string                        `json:"snapshot_table_schema"`
	SlotName            string                         `json:"slot_name" validate:"required"`
	PublicationName     string                         `json:"publication_name" validate:"required"`
	BinaryHandlingMode  string                         `json:"binary_handling_mode" validate:"omitempty,oneof=bytes base64 base64-url-safe hex"`
	HeartbeatEnabled    bool                           `json:"heartbeat_enabled"`
	HeartbeatInterval   *int                           `json:"heartbeat_interval"`
	HeartbeatSchema     *string                        `json:"heartbeat_schema"`
	HeartbeatTable      *string                        `json:"heartbeat_table"`
	TableHierarchy      map[string]map[string][]string `json:"table_hierarchy" validate:"required"`
}

// SetDefault applies default values to the configuration.
func (c *PostgreSQLSourceCreateConfig) SetDefault() error {
	if c.Port == nil {
		defaultPort := 5432
		c.Port = &defaultPort
	}
	if c.SnapshotTableSchema == nil {
		defaultSchema := "public"
		c.SnapshotTableSchema = &defaultSchema
	}
	if c.BinaryHandlingMode == "" {
		c.BinaryHandlingMode = string(Bytes)
	}

	return nil
}
